File: myfortune/Services/AdManager.swift
import GoogleMobileAds
import UIKit

/// Loads and presents the interstitial ad shown to free (non-premium) users
/// right after they view a fortune result.
///
/// Uses Google's public test ad unit ID for now — swap
/// `interstitialAdUnitID` for the real AdMob ad unit ID before shipping to
/// the App Store (see `docs/admob-setup.md` if present, or the AdMob
/// console for this app's ad unit).
final class AdManager: NSObject {
    static let shared = AdManager()

    /// Google's public test interstitial ad unit ID.
    /// TODO: replace with the real ad unit ID before release.
    private let interstitialAdUnitID = "ca-app-pub-3940256099942544/4411468910"

    private var interstitialAd: GADInterstitialAd?
    private var isLoading = false

    private override init() {
        super.init()
    }

    /// Kicks off loading the next interstitial. Safe to call repeatedly —
    /// no-ops while a load is already in flight or an ad is already cached.
    func preloadInterstitial() {
        guard interstitialAd == nil, !isLoading else { return }
        isLoading = true

        GADInterstitialAd.load(withAdUnitID: interstitialAdUnitID, request: GADRequest()) { [weak self] ad, error in
            guard let self else { return }
            self.isLoading = false

            if let error {
                print("AdManager: failed to load interstitial: \(error.localizedDescription)")
                return
            }

            self.interstitialAd = ad
            self.interstitialAd?.fullScreenContentDelegate = self
        }
    }

    /// Presents the interstitial if one is ready. Free users only — callers
    /// are expected to check `PremiumManager.shared.isPremium` first.
    /// Always triggers a preload for the *next* ad, whether or not one was
    /// available to show right now.
    func showInterstitial(from viewController: UIViewController) {
        defer { preloadInterstitial() }

        guard let interstitialAd else {
            print("AdManager: no interstitial ready to show")
            return
        }

        interstitialAd.present(fromRootViewController: viewController)
    }
}

extension AdManager: GADFullScreenContentDelegate {
    func adDidRecordImpression(_ ad: GADFullScreenPresentingAd) {}

    func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        print("AdManager: failed to present interstitial: \(error.localizedDescription)")
        interstitialAd = nil
    }

    func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        interstitialAd = nil
    }
}

File: android/app/src/main/java/com/yourwish/cometrueaifortune/service/AdManager.kt
package com.yourwish.cometrueaifortune.service

import android.app.Activity
import android.content.Context
import android.util.Log
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback

/**
 * Loads and presents the interstitial ad shown to free (non-premium) users
 * right after they view a fortune result. Mirrors myfortune/Services/AdManager.swift.
 *
 * Uses Google's public test interstitial ad unit ID for now — swap
 * [INTERSTITIAL_AD_UNIT_ID] for the real AdMob ad unit ID before shipping to
 * the Play Store (see docs/android-setup.md).
 */
class AdManager private constructor() {

    private var interstitialAd: InterstitialAd? = null
    private var isLoading = false

    /** Kicks off loading the next interstitial. Safe to call repeatedly — no-ops
     *  while a load is already in flight or an ad is already cached. */
    fun preloadInterstitial(context: Context) {
        if (interstitialAd != null || isLoading) return
        isLoading = true

        InterstitialAd.load(
            context,
            INTERSTITIAL_AD_UNIT_ID,
            AdRequest.Builder().build(),
            object : InterstitialAdLoadCallback() {
                override fun onAdLoaded(ad: InterstitialAd) {
                    isLoading = false
                    interstitialAd = ad
                    ad.fullScreenContentCallback = object : FullScreenContentCallback() {
                        override fun onAdFailedToShowFullScreenContent(adError: AdError) {
                            Log.w(TAG, "Failed to present interstitial: ${adError.message}")
                            interstitialAd = null
                        }

                        override fun onAdDismissedFullScreenContent() {
                            interstitialAd = null
                        }
                    }
                }

                override fun onAdFailedToLoad(error: LoadAdError) {
                    isLoading = false
                    Log.w(TAG, "Failed to load interstitial: ${error.message}")
                }
            }
        )
    }

    /** Presents the interstitial if one is ready. Free users only — callers are
     *  expected to check [PremiumManager.isPremium] first. Always triggers a
     *  preload for the *next* ad, whether or not one was available to show now. */
    fun showInterstitial(activity: Activity) {
        val ad = interstitialAd
        if (ad == null) {
            Log.d(TAG, "No interstitial ready to show")
        } else {
            ad.show(activity)
        }
        preloadInterstitial(activity.applicationContext)
    }

    companion object {
        private const val TAG = "AdManager"

        // Google's public test interstitial ad unit ID.
        // TODO: replace with the real ad unit ID before release.
        private const val INTERSTITIAL_AD_UNIT_ID = "ca-app-pub-3940256099942544/1033173712"

        @Volatile
        private var instance: AdManager? = null

        val shared: AdManager
            get() = instance ?: synchronized(this) {
                instance ?: AdManager().also { instance = it }
            }
    }
}
